### PR TITLE
:+1: add `AsyncDisposable` to the return type of `Entrypoint`

### PR DESCRIPTION
**Pull request by Milly**
Implements https://github.com/vim-denops/denops.vim/issues/349
Refs https://github.com/vim-denops/denops.vim/pull/385

## Summary by CodeRabbit

- **New Features**
  - Enhanced plugin system to support asynchronous resource disposal upon unloading, improving the management of resources and stability.

### CODE REVIEW

**Comment by Milly** [comment]
@coderabbitai review

File: denops.ts [hunk 1]
```diff
@@ -156,5 +156,23 @@ export interface Denops {
...
+ *       // Dispose resources...
+ *     }
+ *   }
+ * }
+ * ```
  */
-export type Entrypoint = (denops: Denops) => void | Promise<void>;
+export type Entrypoint = (
+  denops: Denops,
+) => void | AsyncDisposable | Promise<void | AsyncDisposable>;
```
[the remark this answers is not in the document]
Comment (Milly) [reply denops.ts:178]: @lambdalisue I think this is a good suggestion. Would you accept this change?

**Review by lambdalisue — APPROVED** [review]

File: denops.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (lambdalisue) [reply denops.ts:178]: > @lambdalisue I think this is a good suggestion. Would you accept this change?

OK.

[the remark this answers is not in the document]
Comment (Milly) [reply denops.ts:178]: `deno test --doc` failed if replace `void` to `undefined`.
```
error: TS2322 [ERROR]: Type '(denops: Denops) => void' is not assignable to type 'Entrypoint'.
  Type 'void' is not assignable to type 'AsyncDisposable | Promise<AsyncDisposable | undefined> | undefined'.
export const main: Entrypoint = (denops) => {
             ~~~~
    at file:///D:/work/vim/deno-denops-core/denops.ts$152-160.ts:3:14
```

If we replace the return type to `undefined`, will not be able to omit `return`.

[the remark this answers is not in the document]
Comment (Milly) [reply denops.ts:178]: @coderabbitai
`deno test --doc` failed if replace `void` to `undefined`.
```
error: TS2322 [ERROR]: Type '(denops: Denops) => void' is not assignable to type 'Entrypoint'.
  Type 'void' is not assignable to type 'AsyncDisposable | Promise<AsyncDisposable | undefined> | undefined'.
export const main: Entrypoint = (denops) => {
             ~~~~
    at file:///D:/work/vim/deno-denops-core/denops.ts$152-160.ts:3:14
```

If we replace the return type to `undefined`, will not be able to omit `return`.

[the remark this answers is not in the document]
Comment (lambdalisue) [reply denops.ts:178]: Then we should keep void

**Pull request merged**